### PR TITLE
Nil type check implementation

### PR DESCRIPTION
**Pull request by alehander92**

### CODE REVIEW

File: compiler/treetab.nim [hunk 1]
```diff
@@ -10,18 +10,19 @@
 # Implements a table from trees to trees. Does structural equivalence checking.
 
 import
-  hashes, ast, astalgo, types
+  hashes, ast, astalgo, types, renderer
```
Comment (Araq) [inline compiler/treetab.nim]: Debug leftover.

File: compiler/varpartitions_symbol.nim [hunk 2]
```diff
@@ -0,0 +1,1003 @@
...
+## The algorithm is a reinvention / variation of Steensgaard's
+## algorithm.
+## The used data structure is "union find" with path compression.
+
+import ast, types, lineinfos, options, msgs, renderer
+from trees import getMagic, whichPragma
+from wordrecg import wNoSideEffect
+from isolation_check import canAlias
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (Araq) [inline compiler/varpartitions_symbol.nim]: seriously? ;-)

File: compiler/nilcheck.nim [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+
+proc checkBranch(n, ctx, map): Check =
+  result = check(n, ctx, map)
+
+
+# Faith!
+
+proc check(n: PNode, ctx: NilCheckerContext, map: NilMap): Check =
+  if map.isNil:
+    localError(ctx.config, n.info, "map is nil: something went wrong in nilcheck")
```
Comment (Araq) [inline compiler/nilcheck.nim]: Instead use ``internalError``.

File: compiler/nilcheck.nim [hunk 4]
```diff
@@ -0,0 +1,928 @@
...
+    quit 1
```
Comment (Araq) [inline compiler/nilcheck.nim]: No ``quit`` here.

File: compiler/nilcheck.nim [hunk 5]
```diff
@@ -0,0 +1,928 @@
...
+
+  map.store(context.graph(resultId), if not s.typ[0].isNil and s.typ[0].kind == tyRef: Nil else: Safe, TResult, s.ast.info)
+    
+  # echo "checking ", s.name.s, " ", filename
+
+  let res = check(body, context, map)
+  let resultGraphIndex = context.graph(resultId)
+  if res.nilability == Safe and (not res.map.history.hasKey(resultGraphIndex) or res.map.history[resultGraphIndex].len <= 1):
+    res.map.store(resultGraphIndex, Safe, TAssign, s.ast.info)
+  if not s.typ[0].isNil and s.typ[0].kind == tyRef and tfNotNil in s.typ[0].flags:
```
Comment (Araq) [inline compiler/nilcheck.nim:1380]: What if it's not a ``tyRef`` but a tuple containing a ref?

File: compiler/varpartitions_symbol.nim (hunk 2 above)
Comment (alehander92) [reply compiler/varpartitions_symbol.nim]: ugh I forgot to make sure it's removed: I rm-ed it but probably i add-ed again

File: compiler/treetab.nim (hunk 1 above)
Comment (alehander92) [reply compiler/treetab.nim]: ok

File: compiler/nilcheck.nim (hunk 4 above)
Comment (alehander92) [reply compiler/nilcheck.nim]: should I just assert?

Comment (alehander92) [reply compiler/nilcheck.nim]: or just wait and use NilMap not nil later

File: compiler/nilcheck.nim (hunk 5 above)
Comment (alehander92) [reply compiler/nilcheck.nim:1380]: good point: I will think

Comment (alehander92) [reply compiler/nilcheck.nim:1380]: I am starting a commented out test for that: currently i think we should just warn for uninitialized exprs requiring non nilable and for assigning a nilable element to non-nilable (the second one seems to be sometimes happening already in the compiler)

Comment (alehander92) [reply compiler/nilcheck.nim:1380]: but i'll wait for tomorrow

File: compiler/nilcheck.nim (hunk 3 above)
Comment (alehander92) [reply compiler/nilcheck.nim]: for now i'll do so

File: compiler/nilcheck.nim [hunk 6]
```diff
@@ -0,0 +1,1433 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+import ast, renderer, intsets, tables, msgs, options, lineinfos, strformat, idents, treetab, hashes
```
Comment (alehander92) [inline compiler/nilcheck.nim:10]: i am adding a comment so i can link

**Comment by alehander92** [comment]
Hey @Araq do I need to do something right now? How to proceed, there are many possible improvements here

**Comment by alehander92** [comment]
@Araq sorry, I was too lazy to write a comment here: this is ready for review since some time, but not sure what's important to fix

**Comment by Araq** [comment]
Merging this now to keep it from rotting, but will create a follow-up PR.

**Pull request merged**